crates/rpc/ipc/src/server/connection.rs

```diff
@@ -113,3 +116,80 @@ where
         self.project().0.poll_close(cx)
     }
 }
+
+/// Drives an [IpcConn] forward.
+///
+/// This forwards received requests from the connection to the service and sends responses to the
+/// connection.
+///
+/// This future terminates when the connection is closed.
+#[pin_project::pin_project]
+#[must_use = "futures do nothing unless you `.await` or poll them"]
+pub(crate) struct IpcConnDriver<T, S, Fut> {
+    #[pin]
+    pub(crate) conn: IpcConn<JsonRpcStream<T>>,
+    pub(crate) service: S,
+    #[pin]
+    pub(crate) pending_calls: FuturesUnordered<Fut>,
+    pub(crate) items: VecDeque<String>,
+}
+
+impl<T, S, Fut> IpcConnDriver<T, S, Fut> {
+    /// Add a new item to the send queue.
+    pub(crate) fn push_back(&mut self, item: String) {
+        self.items.push_back(item);
+    }
+}
+
+impl<T, S> Future for IpcConnDriver<T, S, S::Future>
```

Comment:
I've seen you write this pattern so many times at this point...do we have any great blogpost which explains this? would be valuable to point contributors towards

Reply:
> do we have any great blogpost

not really, it is mostly dependent on the specific use case, and how the data flows here